Support generating loadable types in serialized function when package-cmo is enabled.

elsh opened the pull request:
To support generating loadable types in a serialized function, return maximal resilience expansion in SILFunction::getResilienceExpansion() if package serialization is enabled for modules built resiliently. 
This allows aggregate types to be generated as loadable SIL types which otherwise are address-only
in a serialized function. During type lowering, opaque flag setting is also skipped if package serialization 
is enabled.

Resolves rdar://127400743

elsh commented:
@swift-ci smoke test

elsh commented:
@swift-ci smoke test

aschwaighofer commented:
Hmm, I think we are working around the fact that type lowering does not correctly answer the question whether a type is loadable in the context we are looking at.

Today, type lowering takes the type expansion context (resilience expansion, etc) into account. I think we need to add another context item: are we lowering for abstraction purposes (i.e. in SILGen to decide the ABI of function signatures and so parameters need the right abstraction) or are we later in the pipeline (in which case we should check whether we can ignore the resilience expansion if `experimental-allow-non-resilient-access`/`experimental-package-bypass-resilience` is enabled)

elsh commented:
@swift-ci smoke test

zoecarver commented:
It seems like the resilience of a type shouldn't change throughout the compilation. If resilience is disabled for package types, that should be represented throughout the whole compilation, right? The `|| isContextPostAbstraction`s seem to be compensating for the fact that type lowering doesn't represent types with bypassed resilience correctly. Can we just teach type lowering about package-bypass-resilience?

elsh commented:
@swift-ci smoke test

slavapestov commented on lib/SIL/IR/SILType.cpp:
```diff
@@ -444,6 +444,10 @@ bool SILType::isLoadableOrOpaque(const SILFunction &F) const {
 }
 
+bool SILType::isLoadable(const SILFunction &F) const {
+  return !isAddressOnly(F) || F.getTypeExpansionContext().isPostAbstraction();
```
This is not the right check. The type might still be address-only for other reasons, you don't want to say that _all_ types are loadable here.

Also, `T.isAddressOnly(F)` should be true if and only if `!T.isLoadable(F)`. This violates that invariant.

slavapestov replied:
You probably want to make your change somewhere near here: https://github.com/apple/swift/blob/ca91d04968af4cebccfa0259ee1322dcaceea8f2/lib/SIL/IR/TypeLowering.cpp#L2337-L2341

elsh committed:
To support serializing functions containing loadable types in a resiliently built
module when package serialization is enabled, return maximal resilience expansion
in SILFunction::getResilienceExpansion(). This allows aggregate types to be generated
as loadable SIL types which otherwise are address-only in a serialized function.
During type lowering, opaque flag setting is also skipped if package serialization
is enabled.

Resolves rdar://127400743

elsh commented:
@swift-ci smoke test

zoecarver commented:
> Resilience does not change. But the context in which we are ask to "expand" a type changes.

I am still struggling with this a bit. Maybe I still don't understand these APIs, but it seems like one of these APIs is being used incorrectly or is implemented incorrectly.

Focusing on the overload of `isResilient` that _does_ take a `ResilienceExpansion`: you're saying this overload should not be used to determine if a type is resilient? If so, maybe we can document that in a comment or even rename it? Regardless, it does seem to be used to determine if a type should be resilient, for example, in IRGen in `TypeConverter::convertStructType` we call `isResilient(D, ResilienceExpansion::Maximal)` which will return `false` for public struct types and influence how those types are lowered. 

This is a huge problem for codegen, if we compile your above example on main:

```
// Core.swift
public struct X { }

public func test() -> X {
	X()
}

// Main.swift
import Core

@inlinable
public func caller() -> X {
  test()
}

// Client.swift
import Main

public func program() {
  _ = caller()
}
```

This produces the following codegen:
```
define swiftcc void @"$s6client7programyyF"() {
entry:
  tail call swiftcc void @"$s4Core4testAA1XVyF"(ptr noalias nocapture sret(%T4Core1XV) undef)
  %0 = tail call ptr @"$s4Core1XVWOh"(ptr undef)
  ret void
}
```

Those undefs are because we are trying to turn an empty explosion into a resilient type, because we now disagree on a type's resilience outside of a package resilience domain. 

Concretely, my suggestions here are:
1. To update type lowering to understand this new flag (by updating `handleResilience`) 
2. Update all overloads of `isResilient` to agree and constrain them to only return false for package decls (today they return false when this flag is on even for public decls).

elsh commented:
> Those undefs are because we are trying to turn an empty explosion into a resilient type, because we now disagree on a type's resilience outside of a package resilience domain.
>
> Concretely, my suggestions here are:
> 
> 1. To update type lowering to understand this new flag (by updating `handleResilience`)
> 2. Update all overloads of `isResilient` to agree and constrain them to only return false for package decls (today they return false when this flag is on even for public decls).

As discussed offline, clients that explicitly opt in to bypass resilience checks to Core types have access to the type layout and are required to be built together with Core, while clients that don't opt in should have indirect access.  This is why `isResilient()` that determines resilience of the definition site needs to be unaltered. Altering will also have a negative impact on what's generated in .swiftinterface where types are expected be resilient.

elsh commented:
The latest updates in this PR no longer involve changes to TypeExpansionContext.

aschwaighofer commented:
Looks good. I assume we remove the serialized flag upon deserialization? Otherwise, we might run into issues with sil-verify-all.

elsh commented:
> Looks good. I assume we remove the serialized flag upon deserialization? Otherwise, we might run into issues with sil-verify-all.

Are you referring to serializePackageEnabled bit? The flag is still there. The removed part in checkResilience is no longer hit because when package serialization is enabled, the returned resilience expansion is now maximal.

aschwaighofer commented:
I referred to the client module that imports the serialized sil. Will that also be compiled with allow-non-resilient-package access?

aschwaighofer commented:
The override of ‘getResilienceExpansion’ will only work in modules that are compiled with ’experimental-allow-non-resilient-access’?

Once you deserialize the function into another module the override will only work if that importing module was also compiled with ‘experimental-allow-non-resilient-access’ or that module has removed the serialize flag upon importing?
